File: 7.cpp
// 根据二叉树先序和中序遍历构建二叉树
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * };
 */
TreeNode* ConstructTree(int* startPreorder, int* endPreorder,
                        int* startInorder, int* endInorder){
    TreeNode* root = new TreeNode(*startPreorder);
    if(startPreorder == endPreorder){
        if(startInorder == endInorder &&
           *startInorder == *startPreorder)
            return root;
        else
            throw std::exception();

    }
    int* rootInorder = startInorder;
    while(rootInorder < endInorder && *rootInorder != *startPreorder){
        rootInorder ++;
    }
    if(rootInorder == endInorder && *rootInorder != *startPreorder){
        throw std::exception();
    }
    int leftLength = rootInorder - startInorder;
    if(startInorder - rootInorder < 0)
        root -> left = ConstructTree(startPreorder + 1, startPreorder + leftLength,
                                     startInorder, startInorder + leftLength - 1);
    if(endInorder - rootInorder > 0)
        root -> right = ConstructTree(startPreorder + leftLength + 1, endPreorder,
                                      startInorder + leftLength + 1, endInorder);
    return root;
}
    

TreeNode* Construct(int* preorder, int* inorder, int length){
    if(preorder == nullptr || inorder == nullptr || length <= 0)
        return nullptr;
    return ConstructTree(preorder, preorder + length - 1,
                  inorder, inorder + length - 1);
}

File: 7.java
/*
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
     
    TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

}
*/

public class Solution {

    public TreeNode reConstructBinaryTree(int[] pre, int[] in) {

        if (pre == null || in == null) {
            return null;
        }

        return reConstructBinaryTree(pre, 0, pre.length - 1, in, 0, in.length - 1);
    }
    
    private TreeNode reConstructBinaryTree
            (int[] pre, int preL, int preR, int[] in, int inL, int inR) {

        if (preL > preR || inL > inR)
            return null;

        int root = pre[preL];
        int left_num = 0;

        for (left_num = 0; in[left_num + inL] != root; ++ left_num) {
            if (left_num + inL >= inR) {
                return null;
            }
        }

        TreeNode node = new TreeNode(pre[preL]);
        node.left = reConstructBinaryTree(pre, preL + 1, preL + left_num, in, inL, inL + left_num - 1);
        node.right = reConstructBinaryTree
              (pre, preL + left_num + 1, preR, in, inL + left_num + 1, inR);

        return node;
        
    }
   
}
